docs/dev/guide/custom-aa-tutorial.md

```diff
@@ -0,0 +1,637 @@
+}
+```
+
+### Transaction validation
+
+Let's implement the validation process. It is responsible for validating the signature of the transaction and incrementing the nonce. Note, that there are some limitation on what this method is allowed to do. You can read more about them [here](../zksync-v2/aa.md#limitations-of-the-verification-step).
+
+To increment the nonce, you should use the `incrementNonceIfEquals` method of the `NONCE_HOLDER_SYSTEM_CONTRACT` system contract. It accepts the nonce of the transaction and checks whether the nonce is the same as the provided one. If not, the transaction reverts. Otherwise, the nonce is increased.
+
+Even though the requirements above allow the accounts to touch only their storage slots, accessing your nonce in the `NONCE_HOLDER_SYSTEM_CONTRACT` is a whitelisted case, since it behaves in the same way as your storage, it just happened to be in another contract. To call the `NONCE_HOLDER_SYSTEM_CONTRACT`, you should add the folloing import:
+
+```solidity
+import '@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol';
+```
+
+The `TransactionHelper` library (already imported in the example above) can be used to get the hash of the transaction that should be signed. You can also implement your own signature scheme and use different a commitment for the transaction to sign, but in this example, we will use the hashes provided by this library.
+
+Using the `TransactionHelper` library:
+
+```solidity
+using TransactionHelper for Transaction;
+```
+
+Now we can implement the `_validateTransaction` method:
+
+```solidity
+function _validateTransaction(Transaction calldata _transaction) internal {
+    // Incrementing the nonce of the account.
+    // Note, that reserved[0] by convention is currently equal to the nonce passed in the transaction
+    NONCE_HOLDER_SYSTEM_CONTRACT.incrementNonceIfEquals(_transaction.reserved[0]);
+    bytes32 txHash = _transaction.encodeHash();
+
+    require(isValidSignature(txHash, _transaction.signature) == EIP1271_SUCCESS_RETURN_VALUE);
+}
+```
+
+### Transaction execution
+
+The most basic implementation of the transaction execution is quite straightforward:
+
+```solidity
+function _executeTransaction(Transaction calldata _transaction) internal {
+    uint256 to = _transaction.to;
+    // By convention, the `reserved[1]` field is msg.value
+    uint256 value = _transaction.reserved[1];
+    bytes memory data = _transaction.data;
+
+    bool success;
+    assembly {
+        success := call(gas(), to, value, add(data, 0x20), mload(data), 0, 0)
+    }
+
+    // Needed for the transaction to be correctly processed by the server.
+    require(success);
+}
+```
+
+Note, that whether the operator will consider the transaction successful will depend only on whether or not the call to `executeTransactions` was successful. So, it is highly recommended to put `require(success)` for the transaction, so that the users could get the best UX.
+
+### Full code for the AA
```

Comment:
```suggestion
### Full code for AA

Reply:
I think it is better to put the `Full code for AA` here.